File: Source/ARTPushDeviceRegistrations.m
#import "ARTPushDeviceRegistrations+Private.h"
#import "ARTHttp.h"
#import "ARTInternalLog.h"
#import "ARTPaginatedResult+Private.h"
#import "ARTDeviceDetails.h"
#import "ARTDevicePushDetails.h"
#import "ARTDeviceIdentityTokenDetails.h"
#import "ARTClientOptions.h"
#import "ARTEncoder.h"
#import "ARTRest+Private.h"
#import "ARTLocalDevice.h"
#import "ARTNSMutableRequest+ARTPush.h"
#import "ARTGCD.h"

@implementation ARTPushDeviceRegistrations {
    ARTQueuedDealloc *_dealloc;
}

- (instancetype)initWithInternal:(ARTPushDeviceRegistrationsInternal *)internal queuedDealloc:(ARTQueuedDealloc *)dealloc {
    self = [super init];
    if (self) {
        _internal = internal;
        _dealloc = dealloc;
    }
    return self;
}

- (void)save:(ARTDeviceDetails *)deviceDetails callback:(ARTCallback)callback {
    [_internal save:deviceDetails wrapperSDKAgents:nil callback:callback];
}

- (void)get:(ARTDeviceId *)deviceId callback:(void (^)(ARTDeviceDetails *_Nullable,  ARTErrorInfo *_Nullable))callback {
    [_internal get:deviceId wrapperSDKAgents:nil callback:callback];
}

- (void)list:(NSStringDictionary *)params callback:(ARTPaginatedDeviceDetailsCallback)callback {
    [_internal list:params wrapperSDKAgents:nil callback:callback];
}

- (void)remove:(NSString *)deviceId callback:(ARTCallback)callback {
    [_internal remove:deviceId wrapperSDKAgents:nil callback:callback];
}

- (void)removeWhere:(NSStringDictionary *)params callback:(ARTCallback)callback {
    [_internal removeWhere:params wrapperSDKAgents:nil callback:callback];
}

@end

@implementation ARTPushDeviceRegistrationsInternal {
    __weak ARTRestInternal *_rest; // weak because rest owns self
    ARTInternalLog *_logger;
    dispatch_queue_t _queue;
    dispatch_queue_t _userQueue;
}

- (instancetype)initWithRest:(ARTRestInternal *)rest logger:(ARTInternalLog *)logger {
    if (self = [super init]) {
        _rest = rest;
        _logger = logger;
        _queue = rest.queue;
        _userQueue = rest.userQueue;
    }
    return self;
}

- (void)save:(ARTDeviceDetails *)deviceDetails wrapperSDKAgents:(nullable NSStringDictionary *)wrapperSDKAgents callback:(ARTCallback)callback {
    if (callback) {
        ARTCallback userCallback = callback;
        callback = ^(ARTErrorInfo *error) {
            art_dispatch_async(self->_userQueue, ^{
                userCallback(error);
            });
        };
    }

    #if TARGET_OS_IOS
    ARTLocalDevice *local = _rest.device;
    #else
    ARTLocalDevice *local = nil;
    #endif

art_dispatch_async(_queue, ^{
    NSMutableDictionary<NSString *, NSString *> *params = nil;
    if (self->_rest.options.pushFullWait) {
        params = @{
            @"fullWait": @"true"
        }.mutableCopy;
    }
    NSData *bodyData = [[self->_rest defaultEncoder] encodeDeviceDetails:deviceDetails error:nil];
    NSError *error = nil;
    NSMutableURLRequest *request = [self->_rest buildRequest:@"PUT"
                                                        path:[@"/push/deviceRegistrations" stringByAppendingPathComponent:deviceDetails.id]
                                                     baseUrl:self->_rest.baseUrl
                                                      params:params
                                                        body:bodyData
                                                     headers:nil
                                                       error:&error];
    if (error) {
        if (callback) callback([ARTErrorInfo createFromNSError:error]);
        return;
    }

    [request setDeviceAuthentication:deviceDetails.id localDevice:local logger:self->_logger];

    ARTLogDebug(self->_logger, @"save device with request %@", request);
    [self->_rest executeAblyRequest:request withAuthOption:ARTAuthenticationOn wrapperSDKAgents:wrapperSDKAgents completion:^(NSHTTPURLResponse *response, NSData *data, NSError *error) {
        if (response.statusCode == 200 /*OK*/) {
            NSError *decodeError = nil;
            ARTDeviceDetails *deviceDetails = [[self->_rest defaultEncoder] decodeDeviceDetails:data error:&decodeError];
            if (decodeError) {
                ARTLogDebug(self->_logger, @"%@: decode device failed (%@)", NSStringFromClass(self.class), error.localizedDescription);
                callback([ARTErrorInfo createFromNSError:decodeError]);
            }
            else {
                ARTLogDebug(self->_logger, @"%@: successfully saved device %@", NSStringFromClass(self.class), deviceDetails.id);
                callback(nil);
            }
        }
        else if (error) {
            ARTLogError(self->_logger, @"%@: save device failed (%@)", NSStringFromClass(self.class), error.localizedDescription);
            callback([ARTErrorInfo createFromNSError:error]);
        }
        else {
            ARTLogError(self->_logger, @"%@: save device failed with status code %ld", NSStringFromClass(self.class), (long)response.statusCode);
            NSString *plain = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
            callback([ARTErrorInfo createWithCode:response.statusCode*100 status:response.statusCode message:[plain art_shortString]]);
        }
    }];
});
}

- (void)get:(ARTDeviceId *)deviceId wrapperSDKAgents:(nullable NSStringDictionary *)wrapperSDKAgents callback:(void (^)(ARTDeviceDetails *, ARTErrorInfo *))callback {
    if (callback) {
        void (^userCallback)(ARTDeviceDetails *, ARTErrorInfo *error) = callback;
        callback = ^(ARTDeviceDetails *device, ARTErrorInfo *error) {
            art_dispatch_async(self->_userQueue, ^{
                userCallback(device, error);
            });
        };
    }

    #if TARGET_OS_IOS
    ARTLocalDevice *local = _rest.device;
    #else
    ARTLocalDevice *local = nil;
    #endif

art_dispatch_async(_queue, ^{
    NSError *error = nil;
    NSMutableURLRequest *request = [self->_rest buildRequest:@"GET"
                                                        path:[@"/push/deviceRegistrations" stringByAppendingPathComponent:deviceId]
                                                     baseUrl:self->_rest.baseUrl
                                                      params:nil
                                                        body:nil
                                                     headers:nil
                                                       error:&error];
    if (error) {
        if (callback) callback(nil, [ARTErrorInfo createFromNSError:error]);
        return;
    }
    
    [request setDeviceAuthentication:deviceId localDevice:local logger:self->_logger];

    ARTLogDebug(self->_logger, @"get device with request %@", request);
    [self->_rest executeAblyRequest:request withAuthOption:ARTAuthenticationOn wrapperSDKAgents:wrapperSDKAgents completion:^(NSHTTPURLResponse *response, NSData *data, NSError *error) {
        if (response.statusCode == 200 /*OK*/) {
            NSError *decodeError = nil;
            ARTDeviceDetails *device = [self->_rest.encoders[response.MIMEType] decodeDeviceDetails:data error:&decodeError];
            if (decodeError) {
                ARTLogDebug(self->_logger, @"%@: decode device failed (%@)", NSStringFromClass(self.class), error.localizedDescription);
                callback(nil, [ARTErrorInfo createFromNSError:decodeError]);
            }
            else if (device) {
                ARTLogDebug(self->_logger, @"%@: get device successfully", NSStringFromClass(self.class));
                callback(device, nil);
            }
            else {
                ARTLogDebug(self->_logger, @"%@: get device failed with unknown error", NSStringFromClass(self.class));
                callback(nil, [ARTErrorInfo createUnknownError]);
            }
        }
        else if (error) {
            ARTLogError(self->_logger, @"%@: get device failed (%@)", NSStringFromClass(self.class), error.localizedDescription);
            callback(nil, [ARTErrorInfo createFromNSError:error]);
        }
        else {
            ARTLogError(self->_logger, @"%@: get device failed with status code %ld", NSStringFromClass(self.class), (long)response.statusCode);
            NSString *plain = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
            callback(nil, [ARTErrorInfo createWithCode:response.statusCode*100 status:response.statusCode message:[plain art_shortString]]);
        }
    }];
});
}

- (void)list:(NSStringDictionary *)params wrapperSDKAgents:(nullable NSStringDictionary *)wrapperSDKAgents callback:(ARTPaginatedDeviceDetailsCallback)callback {
    if (callback) {
        void (^userCallback)(ARTPaginatedResult *, ARTErrorInfo *error) = callback;
        callback = ^(ARTPaginatedResult *result, ARTErrorInfo *error) {
            art_dispatch_async(self->_userQueue, ^{
                userCallback(result, error);
            });
        };
    }

art_dispatch_async(_queue, ^{
    NSError *error = nil;
    NSMutableURLRequest *request = [self->_rest buildRequest:@"GET"
                                                        path:@"/push/deviceRegistrations"
                                                     baseUrl:self->_rest.baseUrl
                                                      params:params
                                                        body:nil
                                                     headers:nil
                                                       error:&error];
    if (error) {
        if (callback) callback(nil, [ARTErrorInfo createFromNSError:error]);
        return;
    }

    ARTPaginatedResultResponseProcessor responseProcessor = ^(NSHTTPURLResponse *response, NSData *data, NSError **error) {
        return [self->_rest.encoders[response.MIMEType] decodeDevicesDetails:data error:error];
    };
    [ARTPaginatedResult executePaginated:self->_rest withRequest:request andResponseProcessor:responseProcessor wrapperSDKAgents:wrapperSDKAgents logger:self->_logger callback:callback];
});
}

- (void)remove:(NSString *)deviceId wrapperSDKAgents:(nullable NSStringDictionary *)wrapperSDKAgents callback:(ARTCallback)callback {
    if (callback) {
        ARTCallback userCallback = callback;
        callback = ^(ARTErrorInfo *error) {
            art_dispatch_async(self->_userQueue, ^{
                userCallback(error);
            });
        };
    }

art_dispatch_async(_queue, ^{
    NSMutableDictionary<NSString *, NSString *> *params = nil;
    if (self->_rest.options.pushFullWait) {
        params = @{
            @"fullWait": @"true"
        }.mutableCopy;
    }
    NSError *error = nil;
    NSMutableURLRequest *request = [self->_rest buildRequest:@"DELETE"
                                                        path:[@"/push/deviceRegistrations" stringByAppendingPathComponent:deviceId]
                                                     baseUrl:self->_rest.baseUrl
                                                      params:params
                                                        body:nil
                                                     headers:nil
                                                       error:&error];
    if (error) {
        if (callback) callback([ARTErrorInfo createFromNSError:error]);
        return;
    }

    ARTLogDebug(self->_logger, @"remove device with request %@", request);
    [self->_rest executeAblyRequest:request withAuthOption:ARTAuthenticationOn wrapperSDKAgents:wrapperSDKAgents completion:^(NSHTTPURLResponse *response, NSData *data, NSError *error) {
        if (response.statusCode == 200 /*Ok*/ || response.statusCode == 204 /*not returning any content*/) {
            ARTLogDebug(self->_logger, @"%@: save device successfully", NSStringFromClass(self.class));
            callback(nil);
        }
        else if (error) {
            ARTLogError(self->_logger, @"%@: remove device failed (%@)", NSStringFromClass(self.class), error.localizedDescription);
            callback([ARTErrorInfo createFromNSError:error]);
        }
        else {
            ARTLogError(self->_logger, @"%@: remove device failed with status code %ld", NSStringFromClass(self.class), (long)response.statusCode);
            NSString *plain = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
            callback([ARTErrorInfo createWithCode:response.statusCode*100 status:response.statusCode message:[plain art_shortString]]);
        }
    }];
});
}

- (void)removeWhere:(NSStringDictionary *)params wrapperSDKAgents:(nullable NSStringDictionary *)wrapperSDKAgents callback:(ARTCallback)callback {
    if (callback) {
        ARTCallback userCallback = callback;
        callback = ^(ARTErrorInfo *error) {
            art_dispatch_async(self->_userQueue, ^{
                userCallback(error);
            });
        };
    }

    #if TARGET_OS_IOS
    ARTLocalDevice *local = _rest.device;
    #else
    ARTLocalDevice *local = nil;
    #endif

art_dispatch_async(_queue, ^{
    NSMutableDictionary<NSString *, NSString *> *queryParams = [params mutableCopy];
    if (self->_rest.options.pushFullWait) {
        queryParams[@"fullWait"] = @"true";
    }
    NSError *error = nil;
    NSMutableURLRequest *request = [self->_rest buildRequest:@"DELETE"
                                                        path:@"/push/deviceRegistrations"
                                                     baseUrl:self->_rest.baseUrl
                                                      params:queryParams
                                                        body:nil
                                                     headers:nil
                                                       error:&error];
    if (error) {
        if (callback) callback([ARTErrorInfo createFromNSError:error]);
        return;
    }
    
    [request setDeviceAuthentication:[params objectForKey:@"deviceId"] localDevice:local];

    ARTLogDebug(self->_logger, @"remove devices with request %@", request);
    [self->_rest executeAblyRequest:request withAuthOption:ARTAuthenticationOn wrapperSDKAgents:wrapperSDKAgents completion:^(NSHTTPURLResponse *response, NSData *data, NSError *error) {
        if (response.statusCode == 200 /*Ok*/ || response.statusCode == 204 /*not returning any content*/) {
            ARTLogDebug(self->_logger, @"%@: remove devices successfully", NSStringFromClass(self.class));
            callback(nil);
        }
        else if (error) {
            ARTLogError(self->_logger, @"%@: remove devices failed (%@)", NSStringFromClass(self.class), error.localizedDescription);
            callback([ARTErrorInfo createFromNSError:error]);
        }
        else {
            ARTLogError(self->_logger, @"%@: remove devices failed with status code %ld", NSStringFromClass(self.class), (long)response.statusCode);
            NSString *plain = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
            callback([ARTErrorInfo createWithCode:response.statusCode*100 status:response.statusCode message:[plain art_shortString]]);
        }
    }];
});
}

@end

File: Sources/AblySwift/ARTPushDeviceRegistrations.swift
import Foundation

/**
 The protocol upon which the `ARTPushDeviceRegistrations` is implemented.
 */
// swift-migration: original location ARTPushDeviceRegistrations.h, line 12
public protocol ARTPushDeviceRegistrationsProtocol {
    
    /**
     * Registers or updates a `ARTDeviceDetails` object with Ably.
     *
     * @param deviceDetails The `ARTDeviceDetails` object to create or update.
     * @param callback A success or failure callback function.
     */
    func save(_ deviceDetails: ARTDeviceDetails, callback: @escaping ARTCallback)
    
    /**
     * Retrieves the `ARTDeviceDetails` of a device registered to receive push notifications using its `deviceId`.
     *
     * @param deviceId The unique ID of the device.
     * @param callback A callback for receiving the `ARTDeviceDetails` object.
     */
    func get(_ deviceId: String, callback: @escaping (ARTDeviceDetails?, ARTErrorInfo?) -> Void)
    
    /**
     * Retrieves all devices matching the filter `params` provided. Returns a `ARTPaginatedResult` object, containing an array of `ARTDeviceDetails` objects.
     *
     * @param params An object containing key-value pairs to filter devices by. Can contain `clientId`, `deviceId` and a `limit` on the number of devices returned, up to 1,000.
     * @param callback A callback for retriving an `ARTPaginatedResult` object with an array of `ARTDeviceDetails` objects.
     */
    func list(_ params: [String: String], callback: @escaping ARTPaginatedDeviceDetailsCallback)
    
    /**
     * Removes a device registered to receive push notifications from Ably using its `deviceId`.
     *
     * @param deviceId The unique ID of the device.
     * @param callback A success or failure callback function.
     */
    func remove(_ deviceId: String, callback: @escaping ARTCallback)
    
    /**
     * Removes all devices registered to receive push notifications from Ably matching the filter `params` provided.
     *
     * @param params An object containing key-value pairs to filter devices by. Can contain `clientId` and `deviceId`.
     * @param callback A success or failure callback function.
     */
    func removeWhere(_ params: [String: String], callback: @escaping ARTCallback)
}

/**
 * Enables the management of push notification registrations with Ably.
 *
 * @see See `ARTPushDeviceRegistrationsProtocol` for details.
 */
// swift-migration: original location ARTPushDeviceRegistrations.h, line 65 and ARTPushDeviceRegistrations.m, line 14
public class ARTPushDeviceRegistrations: NSObject, ARTPushDeviceRegistrationsProtocol, @unchecked Sendable {

    // swift-migration: original location ARTPushDeviceRegistrations+Private.h, line 27
    internal let internalInstance: ARTPushDeviceRegistrationsInternal
    
    private let _dealloc: ARTQueuedDealloc
    
    // swift-migration: original location ARTPushDeviceRegistrations.m, line 18
    internal init(internal internalInstance: ARTPushDeviceRegistrationsInternal, queuedDealloc dealloc: ARTQueuedDealloc) {
        self.internalInstance = internalInstance
        self._dealloc = dealloc
        super.init()
    }
    
    // swift-migration: original location ARTPushDeviceRegistrations.m, line 27
    public func save(_ deviceDetails: ARTDeviceDetails, callback: @escaping ARTCallback) {
        internalInstance.save(deviceDetails, wrapperSDKAgents: nil, callback: callback)
    }
    
    // swift-migration: original location ARTPushDeviceRegistrations.m, line 31
    public func get(_ deviceId: String, callback: @escaping (ARTDeviceDetails?, ARTErrorInfo?) -> Void) {
        internalInstance.get(deviceId, wrapperSDKAgents: nil, callback: callback)
    }
    
    // swift-migration: original location ARTPushDeviceRegistrations.m, line 35
    public func list(_ params: [String: String], callback: @escaping ARTPaginatedDeviceDetailsCallback) {
        internalInstance.list(params, wrapperSDKAgents: nil, callback: callback)
    }
    
    // swift-migration: original location ARTPushDeviceRegistrations.m, line 39
    public func remove(_ deviceId: String, callback: @escaping ARTCallback) {
        internalInstance.remove(deviceId, wrapperSDKAgents: nil, callback: callback)
    }
    
    // swift-migration: original location ARTPushDeviceRegistrations.m, line 43
    public func removeWhere(_ params: [String: String], callback: @escaping ARTCallback) {
        internalInstance.removeWhere(params, wrapperSDKAgents: nil, callback: callback)
    }
}

// swift-migration: original location ARTPushDeviceRegistrations+Private.h, line 9 and ARTPushDeviceRegistrations.m, line 49
internal class ARTPushDeviceRegistrationsInternal: NSObject {
    
    private weak var _rest: ARTRestInternal? // weak because rest owns self
    private let _logger: InternalLog
    private let _queue: DispatchQueue
    private let _userQueue: DispatchQueue
    
    // swift-migration: original location ARTPushDeviceRegistrations.m, line 56
    internal init(rest: ARTRestInternal, logger: InternalLog) {
        self._rest = rest
        self._logger = logger
        self._queue = rest.queue
        self._userQueue = rest.userQueue
        super.init()
    }
    
    // swift-migration: original location ARTPushDeviceRegistrations.m, line 66
    internal func save(_ deviceDetails: ARTDeviceDetails, wrapperSDKAgents: [String: String]?, callback: @escaping ARTCallback) {
        let wrappedCallback: ARTCallback = { error in
            self._userQueue.async {
                callback(error)
            }
        }
        
    #if os(iOS)
        let local = _rest?.device
    #else
        let local: ARTLocalDevice? = nil
    #endif
        
        _queue.async {
            guard let rest = self._rest else {
                wrappedCallback(ARTErrorInfo.create(withCode: 0, message: "ARTRest instance is nil"))
                return
            }
            
            let baseURL = URL(string: "/push/deviceRegistrations")!
            let deviceURL = baseURL.appendingPathComponent(deviceDetails.id ?? "")
            var components = URLComponents(url: deviceURL, resolvingAgainstBaseURL: false)!
            if rest.options.pushFullWait {
                components.queryItems = [URLQueryItem(name: "fullWait", value: "true")]
            }
            
            var request = URLRequest(url: components.url!)
            request.httpMethod = "PUT"
            do {
                request.httpBody = try rest.defaultEncoder.encodeDeviceDetails(deviceDetails)
            } catch {
                request.httpBody = nil
            }
            request.setValue(rest.defaultEncoder.mimeType(), forHTTPHeaderField: "Content-Type")
            
            if let deviceId = deviceDetails.id, let localDevice = local,
               let mutableRequest = (request.settingDeviceAuthentication(deviceId, localDevice: localDevice, logger: self._logger) as NSURLRequest).mutableCopy() as? NSMutableURLRequest {
                request = mutableRequest as URLRequest
            }
            
            ARTLogDebug(self._logger, "save device with request \(request)")
            _ = rest.executeRequest(request, withAuthOption: .on, wrapperSDKAgents: wrapperSDKAgents) { response, data, error in
                if let response {
                    if response.statusCode == 200 {
                        if let data = data {
                            do {
                                let deviceDetails = try rest.defaultEncoder.decodeDeviceDetails(data)
                                if let deviceDetails = deviceDetails {
                                    ARTLogDebug(self._logger, "\(String(describing: type(of: self))): successfully saved device \(deviceDetails.id ?? "")")
                                }
                                wrappedCallback(nil)
                            } catch {
                                ARTLogDebug(self._logger, "\(String(describing: type(of: self))): decode device failed (\(error.localizedDescription))")
                                wrappedCallback(ARTErrorInfo.createFromNSError(error))
                            }
                        } else {
                            wrappedCallback(nil)
                        }
                    } else {
                        ARTLogError(self._logger, "\(String(describing: type(of: self))): save device failed with status code \(response.statusCode)")
                        let plain = data.flatMap { String(data: $0, encoding: .utf8) } ?? ""
                        wrappedCallback(ARTErrorInfo.create(withCode: response.statusCode * 100, status: response.statusCode, message: plain.art_shortString))
                    }
                } else if let error = error {
                    ARTLogError(self._logger, "\(String(describing: type(of: self))): save device failed (\(error.localizedDescription))")
                    wrappedCallback(ARTErrorInfo.createFromNSError(error))
                }
            }
        }
    }
    
    // swift-migration: original location ARTPushDeviceRegistrations.m, line 120
    internal func get(_ deviceId: String, wrapperSDKAgents: [String: String]?, callback: @escaping (ARTDeviceDetails?, ARTErrorInfo?) -> Void) {
        let wrappedCallback: (ARTDeviceDetails?, ARTErrorInfo?) -> Void = { device, error in
            self._userQueue.async {
                callback(device, error)
            }
        }
        
    #if os(iOS)
        let local = _rest?.device
    #else
        let local: ARTLocalDevice? = nil
    #endif
        
        _queue.async {
            guard let rest = self._rest else {
                wrappedCallback(nil, ARTErrorInfo.create(withCode: 0, message: "ARTRest instance is nil"))
                return
            }
            
            let baseURL = URL(string: "/push/deviceRegistrations")!
            let deviceURL = baseURL.appendingPathComponent(deviceId)
            var request = URLRequest(url: deviceURL)
            request.httpMethod = "GET"
            
            if let localDevice = local,
               let mutableRequest = (request.settingDeviceAuthentication(deviceId, localDevice: localDevice, logger: self._logger) as NSURLRequest).mutableCopy() as? NSMutableURLRequest {
                request = mutableRequest as URLRequest
            }
            
            ARTLogDebug(self._logger, "get device with request \(request)")
            _ = rest.executeRequest(request, withAuthOption: .on, wrapperSDKAgents: wrapperSDKAgents) { response, data, error in
                if let response {
                    if response.statusCode == 200 {
                        if let data = data,
                           let mimeType = response.mimeType,
                           let encoder = rest.encoders[mimeType] {
                            do {
                                let device = try encoder.decodeDeviceDetails(data)
                                if let device = device {
                                    ARTLogDebug(self._logger, "\(String(describing: type(of: self))): get device successfully")
                                    wrappedCallback(device, nil)
                                } else {
                                    ARTLogDebug(self._logger, "\(String(describing: type(of: self))): get device failed with unknown error")
                                    wrappedCallback(nil, ARTErrorInfo.createUnknownError())
                                }
                            } catch {
                                ARTLogDebug(self._logger, "\(String(describing: type(of: self))): decode device failed (\(error.localizedDescription))")
                                wrappedCallback(nil, ARTErrorInfo.createFromNSError(error))
                            }
                        } else {
                            ARTLogDebug(self._logger, "\(String(describing: type(of: self))): get device failed with unknown error")
                            wrappedCallback(nil, ARTErrorInfo.createUnknownError())
                        }
                    } else {
                        ARTLogError(self._logger, "\(String(describing: type(of: self))): get device failed with status code \(response.statusCode)")
                        let plain = data.flatMap { String(data: $0, encoding: .utf8) } ?? ""
                        wrappedCallback(nil, ARTErrorInfo.create(withCode: response.statusCode * 100, status: response.statusCode, message: plain.art_shortString))
                    }
                } else if let error = error {
                    ARTLogError(self._logger, "\(String(describing: type(of: self))): get device failed (\(error.localizedDescription))")
                    wrappedCallback(nil, ARTErrorInfo.createFromNSError(error))
                }
            }
        }
    }
    
    // swift-migration: original location ARTPushDeviceRegistrations.m, line 172
    internal func list(_ params: [String: String], wrapperSDKAgents: [String: String]?, callback: @escaping ARTPaginatedDeviceDetailsCallback) {
        let wrappedCallback: ARTPaginatedDeviceDetailsCallback = { result, error in
            self._userQueue.async {
                callback(result, error)
            }
        }
        
        _queue.async {
            guard let rest = self._rest else {
                wrappedCallback(nil, ARTErrorInfo.create(withCode: 0, message: "ARTRest instance is nil"))
                return
            }
            
            var components = URLComponents(url: URL(string: "/push/deviceRegistrations")!, resolvingAgainstBaseURL: false)!
            components.queryItems = params.art_asURLQueryItems()
            var request = URLRequest(url: components.url!)
            request.httpMethod = "GET"
            
            // swift-migration: Updated responseProcessor to use throws pattern instead of inout error parameter
            let responseProcessor: ARTPaginatedResultResponseProcessor = { response, data in
                guard let response,
                      let data = data,
                      let mimeType = response.mimeType,
                      let encoder = rest.encoders[mimeType] else {
                    return []
                }
                return try encoder.decodeDevicesDetails(data) ?? []
            }
            
            ARTPaginatedResult.executePaginated(rest, withRequest: request, andResponseProcessor: responseProcessor, wrapperSDKAgents: wrapperSDKAgents, logger: self._logger, callback: wrappedCallback)
        }
    }
    
    // swift-migration: original location ARTPushDeviceRegistrations.m, line 195
    internal func remove(_ deviceId: String, wrapperSDKAgents: [String: String]?, callback: @escaping ARTCallback) {
        let wrappedCallback: ARTCallback = { error in
            self._userQueue.async {
                callback(error)
            }
        }
        
        _queue.async {
            guard let rest = self._rest else {
                wrappedCallback(ARTErrorInfo.create(withCode: 0, message: "ARTRest instance is nil"))
                return
            }
            
            let baseURL = URL(string: "/push/deviceRegistrations")!
            let deviceURL = baseURL.appendingPathComponent(deviceId)
            var components = URLComponents(url: deviceURL, resolvingAgainstBaseURL: false)!
            if rest.options.pushFullWait {
                components.queryItems = [URLQueryItem(name: "fullWait", value: "true")]
            }
            
            var request = URLRequest(url: components.url!)
            request.httpMethod = "DELETE"
            request.setValue(rest.defaultEncoder.mimeType(), forHTTPHeaderField: "Content-Type")
            
            ARTLogDebug(self._logger, "remove device with request \(request)")
            _ = rest.executeRequest(request, withAuthOption: .on, wrapperSDKAgents: wrapperSDKAgents) { response, data, error in
                if let response {
                    if response.statusCode == 200 || response.statusCode == 204 {
                        ARTLogDebug(self._logger, "\(String(describing: type(of: self))): save device successfully")
                        wrappedCallback(nil)
                    } else {
                        ARTLogError(self._logger, "\(String(describing: type(of: self))): remove device failed with status code \(response.statusCode)")
                        let plain = data.flatMap { String(data: $0, encoding: .utf8) } ?? ""
                        wrappedCallback(ARTErrorInfo.create(withCode: response.statusCode * 100, status: response.statusCode, message: plain.art_shortString))
                    }
                } else if let error = error {
                    ARTLogError(self._logger, "\(String(describing: type(of: self))): remove device failed (\(error.localizedDescription))")
                    wrappedCallback(ARTErrorInfo.createFromNSError(error))
                }
            }
        }
    }
    
    // swift-migration: original location ARTPushDeviceRegistrations.m, line 233
    internal func removeWhere(_ params: [String: String], wrapperSDKAgents: [String: String]?, callback: @escaping ARTCallback) {
        let wrappedCallback: ARTCallback = { error in
            self._userQueue.async {
                callback(error)
            }
        }
        
    #if os(iOS)
        let local = _rest?.device
    #else
        let local: ARTLocalDevice? = nil
    #endif
        
        _queue.async {
            guard let rest = self._rest else {
                wrappedCallback(ARTErrorInfo.create(withCode: 0, message: "ARTRest instance is nil"))
                return
            }
            
            var components = URLComponents(url: URL(string: "/push/deviceRegistrations")!, resolvingAgainstBaseURL: false)!
            components.queryItems = params.art_asURLQueryItems()
            if rest.options.pushFullWait {
                let existingItems = components.queryItems ?? []
                components.queryItems = existingItems + [URLQueryItem(name: "fullWait", value: "true")]
            }
            
            var request = URLRequest(url: components.url!)
            request.httpMethod = "DELETE"
            
            if let deviceId = params["deviceId"], let localDevice = local,
               let mutableRequest = (request.settingDeviceAuthentication(deviceId, localDevice: localDevice) as NSURLRequest).mutableCopy() as? NSMutableURLRequest {
                request = mutableRequest as URLRequest
            }
            
            ARTLogDebug(self._logger, "remove devices with request \(request)")
            _ = rest.executeRequest(request, withAuthOption: .on, wrapperSDKAgents: wrapperSDKAgents) { response, data, error in
                if let response {
                    if response.statusCode == 200 || response.statusCode == 204 {
                        ARTLogDebug(self._logger, "\(String(describing: type(of: self))): remove devices successfully")
                        wrappedCallback(nil)
                    } else {
                        ARTLogError(self._logger, "\(String(describing: type(of: self))): remove devices failed with status code \(response.statusCode)")
                        let plain = data.flatMap { String(data: $0, encoding: .utf8) } ?? ""
                        wrappedCallback(ARTErrorInfo.create(withCode: response.statusCode * 100, status: response.statusCode, message: plain.art_shortString))
                    }
                } else if let error = error {
                    ARTLogError(self._logger, "\(String(describing: type(of: self))): remove devices failed (\(error.localizedDescription))")
                    wrappedCallback(ARTErrorInfo.createFromNSError(error))
                }
            }
        }
    }
}
